tests/test_covar2d.py

```diff
@@ -154,3 +156,59 @@ def testGetCWFCoeffsClean(self):
         self.assertTrue(
             np.allclose(results, self.coeff_cwf_clean, atol=utest_tolerance(self.dtype))
         )
+
+    def testGetCWFCoeffsCleanCTF(self):
+        """
+        Test case of clean images (coeff_clean and noise_var=0)
+        while using a non Identity CTF.
+
+        This case may come up when a developer switches between
+        clean and dirty images.
+        """
+
+        coeff_cwf = self.cov2d.get_cwf_coeffs(
+            self.coeff_clean, self.h_ctf_fb, self.h_idx, noise_var=0
+        )
+
+        # Reconstruct images from output of get_cwf_coeffs
+        img_est = self.basis.evaluate(coeff_cwf)
+        # Compare with clean images
+        delta = np.mean(np.square((self.imgs_clean - img_est).asnumpy()))
+        self.assertTrue(delta < 0.02)
+
+    def testGetCWFCoeffsCTFargs(self):
+        """
+        Test we raise when user supplies incorrect CTF arguments
+        """
+
+        with raises(RuntimeError):
+            _ = self.cov2d.get_cwf_coeffs(
+                self.coeff, self.h_ctf_fb, None, noise_var=self.noise_var
+            )
+
+    @parameterized.expand(
+        [(None,), "frobenius_norm", "operator_norm", "soft_threshold", "None"]
+    )
+    def testShrinkers(self, shrinker):
+        """Test all the shrinkers we know about run, and we raise otherwise."""
+
+        if shrinker == "None":
```

Comment:
Maybe another string would be better here (this just gives me flashbacks).

Reply:
This was intentional. I will add another string.

Reply:
I broke these out into two lists (`shrinkers` and `bad_shrinker_inputs`) at the class level so their intent is clearer.

While I was in there I also check the message returned in the exception, and added a note about removing `parameterized` if/when ASPIRE-Python moves away from the TestCase style.